File: dbms/backend/routes/border_patrol.py
import random, string
from flask import Blueprint, request
from database import get_db, row_to_dict, rows_to_list, api_response, api_error

border_patrol_bp = Blueprint('border_patrol', __name__)


def _gen_provisional_id(force):
    code_map = {
        'BSF': 'BSF', 'ITBP': 'ITBP', 'SSB': 'SSB',
        'Assam Rifles': 'AR', 'CISF': 'CISF'
    }
    code = code_map.get(force, 'BP')
    num  = random.randint(1, 999999)
    return f"PROV-{code}-2026-{num:06d}"


@border_patrol_bp.route('/watchlist-check', methods=['POST'])
def watchlist_check():
    data = request.get_json(silent=True) or {}
    passport_no = data.get('passport_no', '').strip()
    name        = data.get('name', '').strip()

    if not passport_no and not name:
        return api_error('passport_no or name required')

    db = get_db()
    try:
        row = None
        if passport_no:
            row = db.execute(
                "SELECT * FROM entities WHERE passport_no=? AND is_blacklist=1",
                (passport_no,)
            ).fetchone()
        if not row and name:
            row = db.execute(
                "SELECT * FROM entities WHERE name LIKE ? AND is_blacklist=1",
                (f'%{name}%',)
            ).fetchone()
    finally:
        db.close()

    if row:
        r = dict(row)
        return api_response(data={
            'is_blacklist':      True,
            'matched_name':      r['name'],
            'blacklist_reason':  r['blacklist_reason'],
            'status':            r['status'],
            'nationality':       r['nationality'],
            'risk_score':        r['risk_score']
        })
    return api_response(data={'is_blacklist': False})


@border_patrol_bp.route('/register-refugee', methods=['POST'])
def register_refugee():
    data = request.get_json(silent=True) or {}
    required = ['name', 'nationality', 'force', 'entry_point', 'assigned_camp',
                'assigned_ngo', 'ngo_message', 'registered_by']
    for f in required:
        if not data.get(f):
            return api_error(f'Field required: {f}')

    if len(data['ngo_message'].strip()) < 50:
        return api_error('NGO message must be at least 50 characters')

    provisional_id = _gen_provisional_id(data['force'])
    entity_id      = f"BMS-REG-{random.randint(10000,99999)}"
    reg_id         = f"REG-{random.randint(100000,999999)}"
    passport_no    = data.get('passport_no') or f"TEMP-{random.randint(1000,9999)}"

    help_tags = ','.join([t for t in [
        'Medical'           if data.get('needs_medical') else '',
        'Shelter'           if data.get('needs_shelter') else '',
        'Legal Aid'         if data.get('needs_legal') else '',
        'Child Protection'  if data.get('needs_child') else '',
        'Education'         if data.get('needs_education') else '',
    ] if t])

    ngo_assignment_id = f"NGA-{random.randint(100000,999999)}"

    db = get_db()
    try:
        # Upsert entity
        db.execute("""
            INSERT OR IGNORE INTO entities
              (id, name, passport_no, nationality, type, entry_point, status,
               risk_score, is_blacklist, last_seen, medical_needs, dob, gender,
               visit_reason, assigned_camp, assigned_ngo, help_tags, officer_notes,
               created_at, updated_at)
            VALUES (?,?,?,?,?,?,?,?,?,datetime('now'),?,?,?,?,?,?,?,?,datetime('now'),datetime('now'))
        """, (
            entity_id, data['name'], passport_no, data['nationality'],
            'Refugee', data['entry_point'], 'Provisional', 10, 0,
            data.get('medical_needs','None'),
            data.get('dob'), data.get('gender'),
            'Seeking Asylum', data['assigned_camp'], data['assigned_ngo'],
            help_tags, data.get('officer_notes','')
        ))

        db.execute("""
            INSERT INTO refugee_registrations
              (id, entity_id, registered_by, force, entry_point, assigned_camp,
               assigned_ngo, help_tags, ngo_message, provisional_id, registration_date, status)
            VALUES (?,?,?,?,?,?,?,?,?,?,datetime('now'),'Active')
        """, (
            reg_id, entity_id, data['registered_by'], data['force'],
            data['entry_point'], data['assigned_camp'], data['assigned_ngo'],
            help_tags, data['ngo_message'], provisional_id
        ))

        db.execute("""
            INSERT INTO ngo_assignments
              (id, refugee_registration_id, ngo_id, ngo_name, message, status, created_at)
            VALUES (?,?,?,?,?,'Pending',datetime('now'))
        """, (
            ngo_assignment_id, reg_id,
            data.get('ngo_id', 'NGO-001'), data['assigned_ngo'],
            data['ngo_message']
        ))

        db.commit()
    except Exception as e:
        db.close()
        return api_error(f'Database error: {str(e)}', 500)
    finally:
        db.close()

    return api_response(data={
        'provisional_id':  provisional_id,
        'registration_id': reg_id,
        'entity_id':       entity_id
    }, message='Refugee registered successfully')


@border_patrol_bp.route('/refugees', methods=['GET'])
def list_refugees():
    force = request.args.get('force', '')
    limit = min(int(request.args.get('limit', 50)), 200)
    offset = int(request.args.get('offset', 0))

    db = get_db()
    try:
        rows = db.execute("""
            SELECT rr.provisional_id, e.name, e.nationality, e.assigned_camp,
                   e.assigned_ngo, rr.status, rr.registration_date, rr.force,
                   rr.entry_point, e.help_tags, e.medical_needs
            FROM refugee_registrations rr
            JOIN entities e ON e.id = rr.entity_id
            WHERE (? = '' OR rr.force = ?)
            ORDER BY rr.registration_date DESC
            LIMIT ? OFFSET ?
        """, (force, force, limit, offset)).fetchall()

        total = db.execute("""
            SELECT COUNT(*) as c FROM refugee_registrations rr
            WHERE (? = '' OR rr.force = ?)
        """, (force, force)).fetchone()['c']
    finally:
        db.close()

    return api_response(data={
        'items': [dict(r) for r in rows],
        'total': total
    })

File: dbms/frontend/js/border-patrol.js
/* border-patrol.js — Force page logic, refugee form, NGO assignment, camp mini-map */

const FORCE_NAME = document.querySelector('meta[name="dbms-force"]')?.content || 'BSF';

// ── Watchlist Check ───────────────────────────────────────────
let _watchlistTimeout = null;

function setupWatchlistCheck() {
  const passportInput = document.getElementById('f-passport');
  const nameInput     = document.getElementById('f-name');
  if (!passportInput && !nameInput) return;

  const check = async () => {
    const passport = passportInput?.value.trim();
    const name     = nameInput?.value.trim();
    if (!passport && !name) return;
    const res = await apiFetch('/api/border-patrol/watchlist-check', {
      method: 'POST',
      body: JSON.stringify({ passport_no: passport, name })
    });
    if (!res.success) return;
    const alertEl  = document.getElementById('watchlist-alert');
    const clearEl  = document.getElementById('watchlist-clear');
    if (res.data.is_blacklist) {
      if (alertEl) {
        alertEl.classList.add('visible');
        document.getElementById('wl-name').textContent   = res.data.matched_name;
        document.getElementById('wl-reason').textContent = res.data.blacklist_reason;
      }
      if (clearEl) clearEl.style.display = 'none';
    } else {
      if (alertEl) alertEl.classList.remove('visible');
      if (clearEl) clearEl.style.display = 'flex';
    }
  };

  [passportInput, nameInput].filter(Boolean).forEach(el => {
    el.addEventListener('blur', () => {
      clearTimeout(_watchlistTimeout);
      _watchlistTimeout = setTimeout(check, 300);
    });
  });
}

// ── NGO Cards ─────────────────────────────────────────────────
let _selectedNGO = null;

async function loadNGOCards() {
  const container = document.getElementById('ngo-cards-container');
  if (!container) return;
  try {
    const res = await fetch(`/api/ngo/list-by-force?force=${encodeURIComponent(FORCE_NAME)}`);
    const json = await res.json();
    if (!json.success) throw new Error('NGO load failed');
    const all  = { [FORCE_NAME]: json.data };
    const ngos = all[FORCE_NAME] || [];
    container.innerHTML = ngos.map(n => `
      <div class="ngo-card" id="ngo-${n.id}" data-ngo-id="${n.id}" data-ngo-name="${n.name}" onclick="selectNGO('${n.id}','${n.name.replace(/'/g,"\\'")}')">
        <div style="display:flex;align-items:center;justify-content:space-between;margin-bottom:6px">
          <span class="ngo-card-title">${n.name}</span>
          <span class="badge badge-${n.type === 'National' ? 'national' : 'regional'}">${n.type}</span>
        </div>
        <p style="font-size:12px;color:var(--color-text-secondary);line-height:1.6;margin-bottom:8px">${n.description}</p>
        <div style="display:flex;gap:6px;flex-wrap:wrap;margin-bottom:6px">${n.specializations.map(s => `<span class="tag-pill">${s}</span>`).join('')}</div>
        <div style="font-size:11px;color:var(--color-text-muted)">${n.contact} — Load: ${n.current_load} cases</div>
      </div>`).join('');
  } catch {
    if (container) container.innerHTML = '<p class="text-muted">NGO data unavailable</p>';
  }
}

function selectNGO(id, name) {
  document.querySelectorAll('.ngo-card').forEach(c => c.classList.remove('selected'));
  const card = document.getElementById(`ngo-${id}`);
  if (card) card.classList.add('selected');
  _selectedNGO = { id, name };
  document.getElementById('f-ngo-selected').value = name;
  document.getElementById('f-ngo-id').value = id;
}

// ── Character counter ─────────────────────────────────────────
function setupCharCounter() {
  const textarea = document.getElementById('f-ngo-message');
  const counter  = document.getElementById('ngo-msg-count');
  if (!textarea || !counter) return;
  textarea.addEventListener('input', () => {
    const n = textarea.value.length;
    counter.textContent = `${n} / 50 minimum`;
    counter.classList.toggle('min-met', n >= 50);
  });
}

// ── Submit registration ───────────────────────────────────────
async function submitRegistration(e) {
  e.preventDefault();
  const form = document.getElementById('refugee-form');
  if (!validateForm(form)) { showToast('Please fill all required fields','error'); return; }

  const ngoMessage = document.getElementById('f-ngo-message').value;
  if (ngoMessage.trim().length < 50) {
    showToast('NGO message must be at least 50 characters','error');
    document.getElementById('f-ngo-message').classList.add('error');
    return;
  }

  const session = getSession();
  const payload = {
    name:           document.getElementById('f-name').value,
    dob:            document.getElementById('f-dob').value,
    gender:         document.getElementById('f-gender').value,
    nationality:    document.getElementById('f-nationality').value,
    entry_point:    document.getElementById('f-entry-point').value,
    assigned_camp:  document.getElementById('f-camp').value,
    assigned_ngo:   document.getElementById('f-ngo-selected').value,
    ngo_id:         document.getElementById('f-ngo-id').value,
    ngo_message:    ngoMessage,
    passport_no:    document.getElementById('f-passport').value,
    officer_notes:  document.getElementById('f-notes').value,
    needs_medical:  document.getElementById('f-need-medical')?.checked,
    needs_shelter:  document.getElementById('f-need-shelter')?.checked,
    needs_legal:    document.getElementById('f-need-legal')?.checked,
    needs_child:    document.getElementById('f-need-child')?.checked,
    needs_education:document.getElementById('f-need-education')?.checked,
    force:          FORCE_NAME,
    registered_by:  session.user_id || 'Officer'
  };

  const submitBtn = document.getElementById('btn-register');
  if (submitBtn) { submitBtn.disabled = true; submitBtn.textContent = 'Registering...'; }

  const res = await apiFetch('/api/border-patrol/register-refugee', {
    method: 'POST', body: JSON.stringify(payload)
  });

  if (submitBtn) { submitBtn.disabled = false; submitBtn.textContent = 'Register Refugee & Notify NGO →'; }

  if (!res.success) { showToast('Registration failed: ' + res.message,'error'); return; }

  const provId = res.data.provisional_id;
  showSuccessModal(provId);
  form.reset();
  document.querySelectorAll('.ngo-card').forEach(c => c.classList.remove('selected'));
  _selectedNGO = null;
}

function showSuccessModal(provId) {
  const qrContainer = `<div id="qr-modal-canvas" style="display:flex;justify-content:center;margin-top:12px"></div>`;
  openModal({
    title: '✓ Refugee Registered Successfully',
    size: 'lg',
    body: `
      <div class="success-modal-prov">
        <div class="prov-label">Provisional ID — Print and give to refugee</div>
        <div class="prov-id" onclick="navigator.clipboard.writeText('${provId}');showToast('Copied!','success')" title="Click to copy" style="cursor:pointer">${provId}</div>
        <div style="font-size:11px;color:var(--color-text-muted);margin-top:6px">Click ID to copy to clipboard</div>
      </div>
      ${qrContainer}
      <p style="font-size:12px;color:var(--color-text-muted);text-align:center;margin-top:12px">The NGO has been notified. Print this QR code and give it to the refugee as their reference.</p>`,
    footer: `
      <button class="btn btn-secondary" onclick="window.print()">Print QR Sheet</button>
      <button class="btn btn-secondary" onclick="closeModal()">Register Another</button>
      <button class="btn btn-primary" onclick="closeModal();loadRefugeeTable()">View All Refugees</button>`
  });
  setTimeout(() => {
    if (typeof QRCode !== 'undefined') {
      QRCode.toCanvas(document.createElement('canvas'), provId, { width: 180 }, (err, canvas) => {
        if (!err) { const el = document.getElementById('qr-modal-canvas'); if (el) el.appendChild(canvas); }
      });
    }
  }, 100);
}

// ── Refugee table ─────────────────────────────────────────────
async function loadRefugeeTable() {
  const tbody = document.getElementById('refugee-tbody');
  if (!tbody) return;
  tbody.innerHTML = '<tr><td colspan="8" style="text-align:center;padding:20px;color:var(--color-text-muted)">Loading...</td></tr>';
  const res = await apiFetch(`/api/border-patrol/refugees?force=${encodeURIComponent(FORCE_NAME)}&limit=100`);
  if (!res.success || !res.data.items.length) {
    tbody.innerHTML = '<tr><td colspan="8" style="text-align:center;padding:24px;color:var(--color-text-muted)">No refugees registered for this force yet.</td></tr>';
    return;
  }
  tbody.innerHTML = res.data.items.map(r => `
    <tr>
      <td class="font-mono">${r.provisional_id}</td>
      <td><strong>${r.name}</strong></td>
      <td>${r.nationality}</td>
      <td>${formatDate(r.registration_date)}</td>
      <td>${r.assigned_camp?.split(',')[0] || '—'}</td>
      <td>${r.assigned_ngo || '—'}</td>
      <td>${statusBadge(r.status)}</td>
      <td>${(r.help_tags||'').split(',').filter(Boolean).map(t=>`<span class="tag-pill" style="font-size:10px">${t.trim()}</span>`).join(' ')}</td>
    </tr>`).join('');
}

// ── Collapsible sections ──────────────────────────────────────
function setupCollapsibles() {
  document.querySelectorAll('.section-header').forEach(header => {
    header.addEventListener('click', () => {
      header.classList.toggle('collapsed');
      const body = header.nextElementSibling;
      if (body) body.classList.toggle('collapsed');
    });
  });
}

document.addEventListener('DOMContentLoaded', () => {
  setupWatchlistCheck();
  setupCharCounter();
  setupCollapsibles();
  loadNGOCards();

  const campsMiniMapEl = document.getElementById('camps-mini-map');
  const forceCamps = FORCE_CAMPS[FORCE_NAME] || [];
  if (campsMiniMapEl && forceCamps.length && typeof initCampMiniMap === 'function') {
    initCampMiniMap('camps-mini-map', forceCamps);
  }

  document.getElementById('refugee-form')?.addEventListener('submit', submitRegistration);
  loadRefugeeTable();

  // Tab switching on refugee records section
  document.querySelectorAll('.tab-item').forEach(tab => {
    tab.addEventListener('click', () => {
      document.querySelectorAll('.tab-item,.tab-content').forEach(el => el.classList.remove('active'));
      tab.classList.add('active');
      const content = document.getElementById(tab.dataset.tab);
      if (content) content.classList.add('active');
    });
  });
});
